### PR TITLE
Add non-ready namespaces to Clusters interface and use in /namespace/ready endpoint

### PR DESCRIPTION
**Pull request by nbroyles**
**What this PR does / why we need it**:

Previously, non-ready namespaces could not be marked as ready without using the `force` flag due to the coordinator ignoring the non-ready namespace when it was initially discovered. This PR ensures that non-ready namespaces are not discarded on discovery so that the coordinator can associate the appropriate dbnode session with the namespace. With the session in place, the coordinator can gracefully check the dbnode for the namespace availablility and set the namespace as ready for writes if the dbnode check is successful.
 
**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note

```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note

```

### CODE REVIEW

File: src/query/api/v1/handler/namespace/ready_test.go [hunk 1]
```diff
@@ -70,7 +70,7 @@ func TestNamespaceReadyHandler(t *testing.T) {
 		"name": "testNamespace",
 	}
 
-	req := httptest.NewRequest("POST", "/namespace/mark_ready",
+	req := httptest.NewRequest("POST", "/namespace/ready",
```
Comment (nbroyles) [inline src/query/api/v1/handler/namespace/ready_test.go:73]: These URLs aren't actually used by the test, but figured I'd correct them anyway.

**Review by robskillington — APPROVED** [review]
LGTM, I assume marking them as ready will make the watch on etcd promote them to ready (on the dynamic cluster)?

File: src/query/storage/m3/dynamic_cluster.go [hunk 2]
```diff
@@ -326,13 +315,27 @@ func (d *dynamicCluster) updateClusterNamespacesWithLock() {
...
 
 	for _, nsMap := range d.namespacesByEtcdCluster {
-		for _, clusterNamespaces := range nsMap.metadataToClusterNamespaces {
+		for md, clusterNamespaces := range nsMap.metadataToClusterNamespaces {
 			for _, clusterNamespace := range clusterNamespaces {
+				status := md.Options().StagingState().Status()
+				// Don't make non-ready namespaces available for read/write in coordinator, but track
+				// them so that we can have the DB session available when we need to check their
+				// readiness in the /namespace/ready check.
+				if status != namespace.ReadyStagingStatus {
```
Comment (wesleyk) [inline src/query/storage/m3/dynamic_cluster.go:330]: is the unknown state intentionally put in here as well?

**Review by wesleyk — APPROVED** [review]

**Comment by nbroyles** [comment]
> LGTM, I assume marking them as ready will make the watch on etcd promote them to ready (on the dynamic cluster)?

Exactly. The update in `ready` will cause the dynamic impl of clusters to update it's internal state subsequently making those namespaces available for us.

File: src/query/storage/m3/dynamic_cluster.go (hunk 2 above)
Comment (nbroyles) [reply src/query/storage/m3/dynamic_cluster.go:330]: Yup. Namespaces created before staging status was added will be `unknown`, but it's OK to promote them to `ready` all the same.

**Pull request merged**